frontend/lib/resolution/Resolver.cpp

```diff
@@ -6450,17 +6450,105 @@ resolveIterTypeWithTag(Resolver& rv,
   return yieldType;
 }
 
-struct ParamRangeInfo {
-  int64_t current;
-  int64_t end;
+struct BaseParamRangeInfo {
   const Type* yieldType = nullptr;
 
   // Optionally specified properties
   int64_t step = 1;
 
-  static optional<ParamRangeInfo> fromBound(Context* context, ResolutionResultByPostorderID& rr, const AstNode* node) {
-    ParamRangeInfo scratch;
+  virtual ~BaseParamRangeInfo() = default;
+
+  virtual bool done() const = 0;
+  virtual QualifiedType advance(Context* context) = 0;
+};
+
+template <typename R>
+static bool populateParamRangeInfoFromBound(R& scratch, optional<typename R::Iter> low, optional<typename R::Iter> hi, Range::OpKind opKind, int step, int numElts) {
+  if (low && hi) {
+    scratch.current = *low;
+    scratch.end = *hi;
+    R::shift(scratch.end, opKind == Range::OPEN_HIGH ? -1 : 0);
+  } else if (low && numElts >= 0) {
+    scratch.end = scratch.current = *low;
+    R::shift(scratch.end, numElts - 1);
+  } else if (hi && numElts >=0) {
+    scratch.end = *hi;
+    R::shift(scratch.end, opKind == Range::OPEN_HIGH ? -1 : 0);
+    scratch.current = scratch.end;
+    R::shift(scratch.current, - (numElts - 1));
+  } else {
+    return false; // error, can't build range
+  }
+
+  scratch.step = step;
+  if (step < 0) {
+    std::swap(scratch.current, scratch.end);
+  }
+
+  return true;
+}
+
+template <typename R>
+static owned<R> paramRangeInfoFromBound(optional<typename R::Iter> low, optional<typename R::Iter> hi, Range::OpKind opKind, int step, int numElts) {
+  auto scratch = toOwned(new R());
+  bool success = populateParamRangeInfoFromBound<R>(*scratch, low, hi, opKind, step, numElts);
+  return success ? std::move(scratch) : nullptr;
+}
+
+struct IntParamRangeInfo : BaseParamRangeInfo {
+  using Iter = int64_t;
+  int64_t current;
+  int64_t end;
+
+  bool done() const override {
+    return (step > 0) ? (current > end) : (current < end);
+  }
+
+  static void shift(int64_t& val, int64_t step) {
```

Comment:
Could this just be a standalone helper function?

Reply:
Also, the name doesn't immediately explain itself, so if a better name can't be found then a brief comment would be nice (as opposed to having to process the logic oneself).